module/VuFind/src/VuFind/Db/Table/Search.php

```diff
@@ -250,34 +244,65 @@ public function saveSearch(
             }
         };
-        foreach ($this->select($callback) as $oldSearch) {
-            // Deminify the old search:
-            $oldSearchMinified = $oldSearch->getSearchObject();
-            $dupSearch = $oldSearchMinified->deminify($manager);
-            // Check first if classes match:
-            if (get_class($dupSearch) != get_class($newSearch)) {
-                continue;
-            }
-            // Check if URLs match:
-            $oldUrl = $dupSearch->getUrlQuery()->getParams();
-            if ($oldUrl == $newUrl) {
-                // Update the old search only if it wasn't saved:
-                if (!$oldSearch->saved) {
-                    $oldSearch->created = date('Y-m-d H:i:s');
-                    // Keep the ID of the old search:
-                    $newSearchMinified->id = $oldSearchMinified->id;
-                    $oldSearch->search_object = serialize($newSearchMinified);
-                    $oldSearch->save();
+        $results = [];
+        foreach ($this->select($callback) as $match) {
+            $minified = $match->getSearchObject();
+            if ($normalized->isEquivalentToMinifiedSearch($minified)) {
+                $results[] = $match;
+                if (count($results) >= $limit) {
+                    break;
                 }
-                // Update the new search from the existing one
-                $newSearch->updateSaveStatus($oldSearch);
-                return $oldSearch;
             }
         }
+        return $results;
+    }
+
+    /**
+     * Add a search into the search table (history)
+     *
+     * @param SearchNormalizer            $normalizer Search manager
+     * @param \VuFind\Search\Base\Results $newSearch  Search to save
+     * @param string                      $sessionId  Current session ID
+     * @param int|null                    $userId     Current user ID
+     *
+     * @return \VuFind\Db\Row\Search
+     */
+    public function saveSearch(
+        SearchNormalizer $normalizer,
+        $newSearch,
+        $sessionId,
+        $userId
+    ) {
+        $normalized = $normalizer->normalizeSearch($newSearch);
+        $duplicates = $this->getSearchRowsMatchingNormalizedSearch(
+            $normalized,
+            $sessionId,
+            $userId,
+            1 // we only need to identify at most one duplicate match
+        );
+        if ($oldSearch = array_shift($duplicates)) {
+            // Update the old search only if it wasn't saved:
+            if (!$oldSearch->saved) {
+                $oldSearch->created = date('Y-m-d H:i:s');
+                // Keep the ID of the old search:
+                $newSearchMinified = $normalized->getMinified();
+                $newSearchMinified->id = $oldSearch->getSearchObject()->id;
+                $oldSearch->search_object = serialize($newSearchMinified);
+                $oldSearch->save();
+            }
+            // Update the new search from the existing one
```

Comment:
Perhaps something like `// Update the search being saved with values from the existing search:` would make this more clear. Or add a mention that it's for the caller's benefit.

Reply:
See commit [15c216b](https://github.com/vufind-org/vufind/pull/2460/commits/15c216b2cfc7c781b8e81a4b8650d30f6c51176e) -- I renamed some variables and rewrote comments in an effort to make things a bit more clear, but I'm open to further suggestions.

Reply:
Looks good to me now.